test/deform/linearized_deform_test.py

```diff
@@ -0,0 +1,289 @@
+    dg0 = hat_deform_grad0(x)
+    dg1 = hat_deform_grad1(x)
+    v0 = vector_field0(x)
+    v1 = vector_field1(x)
+    return dg0 * v0 + dg1 * v1
+
+
+def fixed_templ_adj_hat0(x, **kwargs):
+    return hat_deform_grad0(x) * hat_function(x)
+
+
+def fixed_templ_adj_hat1(x, **kwargs):
+    return hat_deform_grad1(x) * hat_function(x)
+
+
+def inv_deform_hat(x, **kwargs):
+    """Analytic inverse deformation of the hat function.
+
+    The analytic inverse displacement field (x,y) -> - eps * (x+y, 3xy)."""
+    x0 = x[0] - disp_field0(x)
+    x1 = x[1] - disp_field1(x)
+    return hat_function([x0, x1])
+
+
+def exp_div_inv_disp(x, **kwargs):
+    eps = 0.25
+    x0 = -(eps + 3 * eps * x[0])
+    return np.exp(x0)
+
+
+# Test deformation for LinDeformFixedTempl
+def test_fixed_templ():
+    # Define the analytic template as the hat function and its gradient
+    discr_space = odl.uniform_discr(
+        [-1, -1], [1, 1], (30, 30), interp='linear')
+    template = discr_space.element(hat_function)
+
+    # Define the displacement field (x,y) -> eps * (x+y, 3xy)
+    grad_space = odl.ProductSpace(discr_space, discr_space.ndim)
+    disp_field = grad_space.element([disp_field0, disp_field1])
+
+    deform_templ_exact = discr_space.element(deform_hat)
+
+    fixed_templ_op = odl.deform.LinDeformFixedTempl(template)
+    deform_templ_comp_1 = fixed_templ_op(disp_field)
+
+    tmp = (deform_templ_exact - deform_templ_comp_1).norm()
+
+    rlt_err = tmp / deform_templ_comp_1.norm()
+
+    assert rlt_err < 0.03
+
+
+# Test derivative for LinDeformFixedTemplDeriv
+# Define the vector field where the deriative of the fixed template
+# operator is evaluated. This will be the vector field (x,y) -> (x-y, 4xy)
+def test_fixed_templ_deriv():
+    # Define the analytic template as the hat function and its gradient
+    discr_space = odl.uniform_discr(
+        [-1, -1], [1, 1], (30, 30), interp='linear')
+    template = discr_space.element(hat_function)
+
+    # Define the displacement field (x,y) -> eps * (x+y, 3xy)
+    grad_space = odl.ProductSpace(discr_space, discr_space.ndim)
+    disp_field = grad_space.element([disp_field0, disp_field1])
+
+    vector_field = grad_space.element([vector_field0, vector_field1])
+
+    fixed_templ_deriv_exact = discr_space.element(fixed_templ_deriv_hat)
+
+    fixed_templ_deriv_op = odl.deform.LinDeformFixedTemplDeriv(template,
+                                                               disp_field)
+    fixed_templ_deriv_comp_1 = fixed_templ_deriv_op(vector_field)
+
+    fixed_templ_op = odl.deform.LinDeformFixedTempl(template)
+    fixed_templ_op_deriv = fixed_templ_op.derivative(disp_field)
+    fixed_templ_deriv_comp_2 = fixed_templ_op_deriv(vector_field)
+
+    assert all_equal(fixed_templ_deriv_comp_1, fixed_templ_deriv_comp_2)
+
+    tmp = (fixed_templ_deriv_exact - fixed_templ_deriv_comp_1).norm()
+
+    rlt_err = tmp / fixed_templ_deriv_comp_1.norm()
+
+    assert rlt_err < 0.5
```

Comment:
What we need to do here is to compute function `grad(I)(x + v(x)).T u(x)`. 

The used hat function is not differentiable at `abs(x[0]) = r` or `abs(x[1]) = r` . So we need special treatment to obtain the gradient in analytic case. The treatment is different from that in computed case. Additionally, the interpolation also needs to be done in computed case. When the sampling in the `discr_space` is not sufficient, the above two reasons will cause the large error.

I made a test: if the sampling in the `discr_space` is `30*30`, the `rlt_err` is `0.484376663426`; if that is `3000*3000`, the `rlt_err` is `0.0429668195858`.